docs: document kubelet cgroup driver detection from the runtime

marquiz opened the pull request:
Documentation for KEP-4033.

Document the KubeletCgroupDriverFromCRI feature gate. Also, add notes of this feature in parts of the documentation that describe cgroup driver configuration.

Refs:
- [KEP-4033](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4033-group-driver-detection-over-cri) (issue https://github.com/kubernetes/enhancements/issues/4033)
- k/k PRs:
  - https://github.com/kubernetes/kubernetes/pull/118770
  - https://github.com/kubernetes/kubernetes/pull/119441

> **NOTE:** no container runtime supports this new feature, yet, so we cannot refer to specific versions of runtimes (yet)

> **NOTE:** replaces/re-submits #42091 which was unwittingly submitted against the main branch

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -559,6 +560,14 @@ Each feature gate is designed for enabling/disabling a specific feature:
   track of the finished Pods to count towards completion.
 - `KMSv2`: Enables KMS v2 API for encryption at rest. See [Using a KMS Provider for data encryption](/docs/tasks/administer-cluster/kms-provider) for more details.
+- `KubeletCgroupDriverFromCRI`: Enable detection of the kubelet cgroup driver
+  configuration option from the CRI. This feature gate requires the user run a
```
```suggestion
  configuration option from the {{<glossary_tooltip term_id="cri" text="CRI">}}.
  You can use this feature gate on nodes with a kubelet that supports the feature gate
  and where there is a CRI
```

marquiz replied:
Thanks. Changed as suggested

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -559,6 +560,14 @@ Each feature gate is designed for enabling/disabling a specific feature:
+- `KubeletCgroupDriverFromCRI`: Enable detection of the kubelet cgroup driver
+  configuration option from the CRI. This feature gate requires the user run a
+  container runtime that supports the `RuntimeConfig` CRI call. If both CRI and
+  Kubelet support this feature, the kubelet will ignore the cgroupDriver
+  (--cgroup-driver) configuration option. If the container runtime doesn't
```
```suggestion
  kubelet support this feature, the kubelet ignores the `cgroupDriver` configuration
  setting (or deprecated `--cgroup-driver` command line argument).
  If you enable this feature gate and the container runtime doesn't
```

marquiz replied:
👍 Changed as suggested

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -559,6 +560,14 @@ Each feature gate is designed for enabling/disabling a specific feature:
+  container runtime that supports the `RuntimeConfig` CRI call. If both CRI and
+  Kubelet support this feature, the kubelet will ignore the cgroupDriver
+  (--cgroup-driver) configuration option. If the container runtime doesn't
+  support it, the kubelet will fallback to using its cgroupDriver option.
```
```suggestion
  support it, the kubelet falls back to using the driver configured using the `cgroupDriver`
  configuration setting.
```

marquiz replied:
👍 Changed as suggested

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -157,6 +157,13 @@ Starting with v1.22 and later, when creating a cluster with kubeadm, if the user
 {{< /note >}}
 
+{{< note >}}
+Starting with v1.28 and later, with KubeletCgroupDriverFromCRI feature gate
```
```suggestion
For Kubernetes v1.28 and later, with the `KubeletCgroupDriverFromCRI`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
```

marquiz replied:
Thanks. Changed as suggested

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -157,6 +157,13 @@ Starting with v1.22 and later, when creating a cluster with kubeadm, if the user
 
+{{< note >}}
+Starting with v1.28 and later, with KubeletCgroupDriverFromCRI feature gate
+enabled and a container runtime that supports the RuntimeConfig CRI rpc,
```
```suggestion
enabled and a container runtime that supports the `RuntimeConfig` CRI RPC,
```

marquiz replied:
👍 Changed as suggested

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -157,6 +157,13 @@ Starting with v1.22 and later, when creating a cluster with kubeadm, if the user
+{{< note >}}
+Starting with v1.28 and later, with KubeletCgroupDriverFromCRI feature gate
+enabled and a container runtime that supports the RuntimeConfig CRI rpc,
+kubelet will automatically detect the cgroup driver from the runtime and the
+cgroupDriver setting in kubelet configuration is ignored.
```
```suggestion
the kubelet automatically detects the appropriate cgroup driver from the runtime,
and ignores the `cgroupDriver` setting within the kubelet configuration.
```

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -157,6 +157,13 @@ Starting with v1.22 and later, when creating a cluster with kubeadm, if the user
 the `cgroupDriver` field under `KubeletConfiguration`, kubeadm defaults it to `systemd`.
 {{< /note >}}
 
+{{< note >}}
```
I wouldn't use a note here. Use a plain paragraph.

marquiz replied:
OK, made it a plain (non-note) paragraph, as suggested

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -251,6 +258,12 @@ sudo systemctl restart containerd
 When using kubeadm, manually configure the
 [cgroup driver for kubelet](/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver).
 
+{{< note >}}
+Starting with v1.28 and later, kubelet cgroup driver does not need to be
+configured if the KubeletCgroupDriverFromCRI feature gate is enabled and a
+version of containerd that supports the RuntimeConfig CRI rpc is being used.
+{{< /note >}}
```
```suggestion
Starting with v1.28 and later, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
```

or
```suggestion
In Kubernetes {{< skew currentVersion >}}, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
```

marquiz replied:
Thx. Chose the first variant suggested above. I don't have strong opinions on this, but personally (as a user) like to easily get the information when some new feature was introduced.

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -292,6 +305,12 @@ You should also note the changed `conmon_cgroup`, which has to be set to the val
 cgroup driver configuration of the kubelet (usually done via kubeadm) and CRI-O
 in sync.
 
+{{< note >}}
+Starting with v1.28 and later, kubelet cgroup driver does not need to be
+configured if the KubeletCgroupDriverFromCRI feature gate is enabled and a
+version of CRI-O that supports the RuntimeConfig CRI rpc is being used.
+{{< /note >}}
```
```suggestion
Starting with v1.28 and later, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
```

marquiz replied:
Thx. Changed as suggested

sftim commented on content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:
```diff
@@ -38,6 +38,13 @@ In v1.22 and later, if the user does not set the `cgroupDriver` field under `Kub
 kubeadm defaults it to `systemd`.
 {{< /note >}}
 
+{{< note >}}
```
Consider making these be one note with two paragraphs.

sftim commented on content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:
```diff
@@ -38,6 +38,13 @@ In v1.22 and later, if the user does not set the `cgroupDriver` field under `Kub
 {{< /note >}}
 
+{{< note >}}
+In v1.28 and later, with KubeletCgroupDriverFromCRI feature gate
+enabled and a container runtime that supports the RuntimeConfig CRI rpc,
+kubelet will automatically detect the cgroup driver from the runtime and the
+kubelet cgroup driver setting does not need to be manually configured.
```
```suggestion
Starting with v1.28 and later, you can enable automatic detection of the
cgroup driver as an alpha feature.
See [systemd cgroup driver](/docs/setup/production-environment/container-runtimes/#systemd-cgroup-driver)
for more details.
```

sftim reviewed (commented):
Thanks

I've added feedback to align this with our [style guide](https://kubernetes.io/docs/contribute/style/style-guide/)

marquiz commented:
> I've added feedback to align this with our [style guide](https://kubernetes.io/docs/contribute/style/style-guide/)

Thanks @sftim for the review. I incorporated the changes you suggested as a single commit.

sftim reviewed (commented):
Looks good for docs.

Can we get a tech review from SIG Node?

marquiz commented:
> Can we get a tech review from SIG Node?

/cc @mikebrow @haircommander

marquiz commented:
/cc mrunalp

mikebrow commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -157,6 +157,12 @@ Starting with v1.22 and later, when creating a cluster with kubeadm, if the user
 the `cgroupDriver` field under `KubeletConfiguration`, kubeadm defaults it to `systemd`.
 {{< /note >}}
 
+For Kubernetes v1.28 and later, with the `KubeletCgroupDriverFromCRI`
```
bit of a tldr issue I believe.. giving the instructions above for kubelet config driven and default if not configured.. then ending with 1.28 alpha feature gate info.

Suggest mentioning the feature gate above in direct association with the kubelet config steps / options. Something like unless the alpha level feature gate is set to automatically retrieve the appropriate cgroup driver from the runtime, additional configuration may be required...

marquiz replied:
I'm a bit cautious on this ("advertising" too much) as this is an alpha feature and there are no released container runtime supporting this, yet. When we get to beta, sure. WDYT @sftim

mikebrow replied:
good point...

sftim replied:
We should document how Kubernetes behaves, even if the compatible runtimes don't exist on release day.

mikebrow commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -251,6 +257,10 @@ sudo systemctl restart containerd
 When using kubeadm, manually configure the
 [cgroup driver for kubelet](/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver).
 
+Starting with v1.28 and later, you can enable automatic detection of the
```
```suggestion
Starting with Kubernetes v1.28 and later, you can enable automatic detection of the
```

sftim replied:
Aside, but also a bit relevant:
I'm slightly wary of saying “and later” about alpha features that we can't prove will graduate to stable.

marquiz replied:
Fair point, have to agree on that. I now changed all of these to `In Kubernetes {{< skew currentVersion >}}...`

ok @sftim @mikebrow ?

mikebrow replied:
given the concern over skew "and later" comments perhaps.. for Kubernetes v1.2.8, you can..

mikebrow commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -292,6 +302,10 @@ You should also note the changed `conmon_cgroup`, which has to be set to the val
 cgroup driver configuration of the kubelet (usually done via kubeadm) and CRI-O
 in sync.
 
+Starting with v1.28 and later, you can enable automatic detection of the
```
```suggestion
Starting with Kubernetes v1.28 and later, you can enable automatic detection of the
```

marquiz replied:
Fixed with `In Kubernetes {{< skew currentVersion >}}`

mikebrow commented on content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:
```diff
@@ -36,6 +36,11 @@ driver of the kubelet.
 In v1.22 and later, if the user does not set the `cgroupDriver` field under `KubeletConfiguration`,
 kubeadm defaults it to `systemd`.
+
+Starting with v1.28 and later, you can enable automatic detection of the
```
see tldr comment

mikebrow reviewed (commented):
see comments, cheers

sftim commented:
Wording looks fine for alpha, what this really needs now is tech signoff.

mikebrow commented:
> Wording looks fine for alpha, what this really needs now is tech signoff.

to that end for containerd .. I'm reviewing @marquiz impl in containerd https://github.com/containerd/containerd/pull/8722, would like to see that merged before providing my signoff. Cheers!

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -157,6 +157,12 @@ Starting with v1.22 and later, when creating a cluster with kubeadm, if the user
 the `cgroupDriver` field under `KubeletConfiguration`, kubeadm defaults it to `systemd`.
 {{< /note >}}
 
+In Kubernetes {{< skew currentVersion >}}, with the `KubeletCgroupDriverFromCRI`
```
Suppose we are in v1.29 now, the above line will be rendered into something like
"In Kubernetes 1.29, with the `KubeletCgroupDriverFromCRI`".
Is this statement still appropriate?
Suppose we are in v1.34 now and the feature gate has been GA'ed and removed,
the above line will be rendered into something like
"In Kubernetes 1.34, with the `KubeletCgroupDriverFromCRI`".
Is this statement still appropriate?

Clearly stating that this feature is in Alpha starting 1.28 is good enough.
Why do we use the `skew` shortcode the make this release number a moving target?

marquiz replied:
I'm indifferent on this, can change it to v1.29. This (`In Kubernetes {{< skew currentVersion >}}`) was something that @sftim suggested in one of his earlier comments so I took it from there.

sftim replied:
>  Suppose we are in v1.34 now and the feature gate has been GA'ed and removed,

When we change Kubernetes, we should remove the docs. If we don't trust ourselves to do that, I would focus on fixing that lack of trust.

If we could trust the release docs process, I think this change would be fine.

marquiz replied:
As a contributor I'm inclined to agree with @sftim about the docs process here. Especially with a KEP involved, the process should ensure that the documentation is changed when the feature is moved to beta and later GA (or removed before graduating further). WDYT @tengqm?

sftim replied:
Per https://github.com/kubernetes/website/pull/42160#issuecomment-1665322332, let's follow the path of least risk to the release, at least for now.

marquiz replied:
K, I changed this now to `In Kubernetes v1.28, ...` (not to `Starting with v1.28 ...` as per https://github.com/kubernetes/website/pull/42160#discussion_r1274196798)

PTAL @sftim @tengqm

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -251,6 +257,10 @@ sudo systemctl restart containerd
 When using kubeadm, manually configure the
 [cgroup driver for kubelet](/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver).
 
+In Kubernetes {{< skew currentVersion >}}, you can enable automatic detection of the
```
Same as above.

In addition to the above confusion, using this `skew` shortcode makes it difficult
to remove contents that are no longer relevant in the future.

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -292,6 +302,10 @@ You should also note the changed `conmon_cgroup`, which has to be set to the val
 cgroup driver configuration of the kubelet (usually done via kubeadm) and CRI-O
 in sync.
 
+In Kubernetes {{< skew currentVersion >}}, you can enable automatic detection of the
```
Same here.

marquiz commented:
ping @mikebrow the containerd implementation (https://github.com/containerd/containerd/pull/8722) is now merged

ping @sftim what's your take on using `{{< skew currentVersion >}}` (https://github.com/kubernetes/website/pull/42160#discussion_r1275119107)?

sftim commented:
> ping @sftim what's your take on using `{{< skew currentVersion >}}` ([#42160 (comment)](https://github.com/kubernetes/website/pull/42160#discussion_r1275119107))?

I very much like like timeless documentation, but I also like having the docs ready for release. I'm reluctantly OK with not using that shortcode if it unblocks approval. After the release, I could propose putting it back.

sftim reviewed (commented):
/lgtm
/approve
